File: deploy/api/field_eff_table.m
% If current folder is .\source\solarpilot_dev\solarpilot\deploy\api then
% the only path needed is the location of the TMY and can be local.
TMY_file="..\climate_files\USA CA Daggett (TMY2).csv";
cp = py.copylot.CoPylot();

% If current folder is any other folder then you have to provide the path 
% where copylot.py and solarpilot.dll are placed, and the TMY needs the full path.
% Uncomment the following lines in that case. 
% CoPylot_path='C:\Users\adelacal\source\solarpilot_dev\solarpilot\deploy\api';
% TMY_file="C:\Users\adelacal\source\solarpilot_dev\solarpilot\deploy\climate_files\USA CA Daggett (TMY2).csv";
% 
% if count(py.sys.path,CoPylot_path) == 0
%     insert(py.sys.path,int32(0),CoPylot_path);
% end
% cp = py.copylot.CoPylot(path=CoPylot_path);

r = cp.data_create();
cp.data_set_string(r, "ambient.0.weather_file", TMY_file);
cp.generate_layout(r);
field = cp.get_layout_info(r);
cp.simulate(r);
flux = cp.get_fluxmap(r);

tic
cp.calculate_optical_efficiency_table(r, py.int(12), py.int(11));
%cp.calculate_optical_efficiency_table(r, py.int(2), py.int(2))  %# For debugging
toc

eff_filename = "./efficiency_table.csv";
cp.save_optical_efficiency_table(r, eff_filename), %# Testing simple table saving (consistent with get_optical_efficiency_table)
cp.save_optical_efficiency_table(r, "./modelica_efficiency_table.txt", "sf_eff_table");  %# Testing Modelica specific table output

sf_eff = cp.get_optical_efficiency_table(r);
cp.data_free(r);

% Read in efficiency table
eff_table=readtable(eff_filename);
ele= eff_table{1,2:end};
azi= eff_table{2:end,1};
eff_data = eff_table{2:end,2:end};

% Check data is consistent between two methods
t1(1)=isequal(azi',round(double(sf_eff.get('azimuth')),4));
t1(2)=isequal(ele,round(double(sf_eff.get('elevation')),4));
t1(3)=isequal(eff_data,round(double(py.numpy.array(sf_eff.get('eff_data'))),6));

FontSize=14;
LineWidth=1.5;
figure()
legInfoAzi=cell(length(azi),1);
hold on
for i=1:length(azi)
    plot(ele,eff_data(i,:),'LineWidth',LineWidth);
    legInfoAzi{i}=num2str(azi(i));
end
hold off
xlabel('Elevation Angle [deg]')
ylabel('Solarfield Efficiency [-]')
set(gca,'FontSize',FontSize)
leg=legend(legInfoAzi,'NumColumns',2,'Location','Best');
title(leg,'Azimuth [deg]')
box on 
grid on
xlim([0,90])

figure()
legInfoEle=cell(length(ele),1);
hold on
for i=1:length(ele)
    plot(azi,eff_data(:,i),'LineWidth',LineWidth);
    legInfoEle{i}=num2str(ele(i));
end
hold off
xlabel('Azimuth Angle [deg]')
ylabel('Solarfield Efficiency [-]')
set(gca,'FontSize',FontSize)
leg=legend(legInfoEle,'NumColumns',2,'Location','Best');
title(leg,'Elevation [deg]')
xlim([0,360])
box on 
grid on

File: deploy/api/field_eff_table.py
import os, csv
from copylot import CoPylot
import time
import numpy as np
import matplotlib.pyplot as plt

plt.rcParams['font.size'] = 14
plt.rcParams['font.sans-serif'] = 'Arial'
plt.rcParams['axes.linewidth'] = 2.5
plt.rcParams['xtick.major.width'] = 1.5
plt.rcParams['ytick.major.width'] = 1.5

eff_filename = "efficiency_table.csv"

cp = CoPylot(debug = False)
print("Process ID: ", os.getpid())
r = cp.data_create()
assert cp.data_set_string(
        r,
        "ambient.0.weather_file",
        "../climate_files/USA CA Daggett (TMY2).csv",
    )
assert cp.generate_layout(r)
# field = cp.get_layout_info(r)
# assert cp.simulate(r)
# flux = cp.get_fluxmap(r)

start = time.time()
assert cp.calculate_optical_efficiency_table(r, 12, 11)
# assert cp.calculate_optical_efficiency_table(r, 2, 2)  # For debugging
end = time.time()
print("Efficiency computation required {:.2f} seconds".format(end-start))

# Testing getting optical efficiency table after calculation
cwd = os.getcwd()
assert cp.save_optical_efficiency_table(r, os.path.join(cwd, eff_filename)) # Testing simple table saving (consistent with get_optical_efficiency_table)
assert cp.save_optical_efficiency_table(r, os.path.join(cwd, "modelica_efficiency_table.txt"), "sf_eff_table")  # Testing Modelica specific table output
sf_eff = cp.get_optical_efficiency_table(r)

start = time.time()
direct = cp.calculate_get_optical_efficiency_table(r, 12, 11) # calculate and get
end = time.time()
print("Direct Calculate and get efficiency required {:.2f} seconds".format(end-start))

assert cp.data_free(r)

# Read in efficiency table
eff_data = []
el = []
azi = []
with open(eff_filename) as f:
    csvReader = csv.reader(f)
    for row in csvReader:
        if el == []:
            el.extend([float(i) for i in row])
            el.pop(0) # pop off the first element
        else:
            azi.append(float(row[0]))
            row.pop(0)
            eff_data.append([float(i) for i in row])

# Check data is consistent between two methods
for a in range(len(sf_eff['azimuth'])):
    assert azi[a] == round(sf_eff['azimuth'][a], 4)
    assert direct['azimuth'][a] == sf_eff['azimuth'][a], 4

for e in range(len(sf_eff['elevation'])):
    assert el[e] == round(sf_eff['elevation'][e], 4)
    assert direct['elevation'][e] == sf_eff['elevation'][e]

for a in range(len(sf_eff['azimuth'])):
    for e in range(len(sf_eff['elevation'])):
        assert eff_data[a][e] == round(sf_eff['eff_data'][a][e], 6)
        assert direct['eff_data'][a][e] == sf_eff['eff_data'][a][e]

# Transpose data for plotting
sf_eff['eff_data_T'] = np.array(sf_eff['eff_data']).T.tolist()

# Plot efficiency data
fig, (ax1, ax2) = plt.subplots(1, 2)
for a in range(len(sf_eff['azimuth'])):
    ax1.plot(sf_eff['elevation'], sf_eff['eff_data'][a], label=str(sf_eff['azimuth'][a]))
ax1.legend(title='Azimuth Angle [deg]')
ax1.set_xlabel('Elevation Angle [deg]')
ax1.set_ylabel('Solarfield Efficiency [-]')

for e in range(len(sf_eff['elevation'])):
    ax2.plot(sf_eff['azimuth'], sf_eff['eff_data_T'][e], label=str(sf_eff['elevation'][e]))
ax2.legend(title='Elevation Angle [deg]')
ax2.set_xlabel('Azimuth Angle [deg]')
ax1.set_ylabel('Solarfield Efficiency [-]')

plt.tight_layout()
plt.show()

